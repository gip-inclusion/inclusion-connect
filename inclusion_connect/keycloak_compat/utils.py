import re


namespace_pattern = re.compile(r"^keycloak_compat_(?P<realm>.+)$")


def realm_from_request(request):
    namespace = request.resolver_match.namespace
    match = namespace_pattern.match(namespace)
    return match.group("realm")
