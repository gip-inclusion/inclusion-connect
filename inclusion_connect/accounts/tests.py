from urllib.parse import quote

from django.contrib import messages
from django.contrib.auth import get_user
from django.core import mail
from django.urls import reverse
from django.utils.encoding import force_bytes
from django.utils.http import urlsafe_base64_encode
from pytest_django.asserts import assertContains, assertRedirects, assertTemplateUsed

from inclusion_connect.accounts.views import PasswordResetView
from inclusion_connect.users.factories import DEFAULT_PASSWORD, UserFactory
from inclusion_connect.users.models import User
from inclusion_connect.utils.urls import add_url_params


def test_login(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:login"), {"next": redirect_url})
    user = UserFactory()
    assert not get_user(client).is_authenticated

    response = client.get(url)
    assertContains(response, "Connexion")
    assertContains(response, "Adresse e-mail")  # Ask for email, not username
    assertContains(response, reverse("accounts:registration"))  # Link to registration page

    response = client.post(url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated


def test_login_failed_bad_email_or_password(client):
    url = add_url_params(reverse("accounts:login"), {"next": "anything"})
    user = UserFactory()
    assert not get_user(client).is_authenticated

    response = client.post(url, data={"email": user.email, "password": "toto"})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated

    response = client.post(url, data={"email": "wrong@email.com", "password": DEFAULT_PASSWORD})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated

    # If user is inactive
    user.is_active = False
    user.save()
    response = client.post(url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated


def test_user_creation(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:registration"), {"next": redirect_url})
    user = UserFactory.build()
    assert not get_user(client).is_authenticated

    response = client.get(url)
    assertContains(response, "Créer un compte")
    assertContains(response, reverse("accounts:login"))  # Link to login page
    assertContains(response, "CGU_20230302.pdf")
    assertContains(response, quote("Politique de confidentialité_20230302.pdf"))

    response = client.post(
        url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
            "terms_accepted": True,
        },
    )
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated
    user = User.objects.get(email=user.email)  # Previous instance was a built factory, so refresh_from_db won't work
    assert user.terms_accepted_at == user.date_joined


def test_user_creation_terms_are_required(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:registration"), {"next": redirect_url})
    user = UserFactory.build()
    assert not get_user(client).is_authenticated

    response = client.post(
        url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
        },
    )
    assertTemplateUsed("registration.html")
    assert "terms_accepted" in response.context["form"].errors


# TODO: Test next url propagation in all the registration process (creation + email validation)


def test_password_reset(client):
    user = UserFactory()

    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:login"), {"next": redirect_url})
    response = client.get(url)
    password_reset_url = reverse("accounts:password_reset")
    assertContains(response, password_reset_url)

    response = client.get(password_reset_url)
    assertTemplateUsed(response, "password_reset.html")

    response = client.post(password_reset_url, data={"email": user.email})
    assertRedirects(response, reverse("accounts:login"))
    assert list(messages.get_messages(response.wsgi_request)) == [
        messages.storage.base.Message(
            messages.SUCCESS,
            "Si un compte existe avec cette adresse e-mail, "
            "vous recevrez un e-mail contenant des instructions pour réinitialiser votre mot de passe.",
        ),
    ]

    # Check sent email
    assert len(mail.outbox) == 1
    uid = urlsafe_base64_encode(force_bytes(user.pk))
    token = PasswordResetView.token_generator.make_token(user)
    password_reset_url = reverse("accounts:password_reset_confirm", args=(uid, token))
    assert password_reset_url in mail.outbox[0].body

    # Change password
    password = "toto"
    response = client.get(password_reset_url)  # retrieve the modified url
    response = client.post(response.url, data={"new_password1": password, "new_password2": password})
    # FIXME: skip redirection to login, go to next url !
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated
