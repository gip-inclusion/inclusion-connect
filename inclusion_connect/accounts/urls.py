from django.urls import path, re_path

from inclusion_connect.accounts import views


app_name = "accounts"

urlpatterns = [
    re_path(r"^login/$", views.LoginView.as_view(), name="login"),
    re_path(r"^register/$", views.RegisterView.as_view(), name="register"),
    re_path(r"^activate/$", views.ActivateAccountView.as_view(), name="activate"),
    re_path(r"^password_reset/$", views.PasswordResetView.as_view(), name="password_reset"),
    path("reset/<uidb64>/<token>/", views.PasswordResetConfirmView.as_view(), name="password_reset_confirm"),
    re_path(r"^accept-terms/$", views.AcceptTermsView.as_view(), name="accept_terms"),
    re_path(r"^my-account/$", views.EditUserInfoView.as_view(), name="edit_user_info"),
    re_path(r"^change-password/$", views.PasswordChangeView.as_view(), name="change_password"),
]
