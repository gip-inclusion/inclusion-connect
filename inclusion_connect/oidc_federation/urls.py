from django.urls import include, re_path

from . import peama


app_name = "oidc_federation"

peama_urlpatterns = [
    re_path(r"^callback/", peama.OIDCAuthenticationCallbackView.as_view(), name="callback"),
    re_path(r"^authenticate/", peama.OIDCAuthenticationRequestView.as_view(), name="init"),
    # There's no logout available on PEAMA
]

urlpatterns = [
    re_path(r"^peama/", include((peama_urlpatterns, "peama"))),
]
