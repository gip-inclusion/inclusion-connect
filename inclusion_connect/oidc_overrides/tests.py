from django.contrib.auth import get_user
from django.urls import reverse
from pytest_django.asserts import assertRedirects

from inclusion_connect.oidc_overrides.factories import ApplicationFactory
from inclusion_connect.oidc_overrides.views import OIDCSessionMixin
from inclusion_connect.users.factories import UserFactory
from inclusion_connect.utils.urls import add_url_params


OIDC_PARAMS = {
    "response_type": "code",
    "client_id": "my_application",
    "redirect_uri": "http://localhost/callback",
    "scope": "openid profile email",
    "state": "state",
    "nonce": "nonce",
}


def test_allow_wildcard_in_redirect_uris():
    application = ApplicationFactory(redirect_uris="http://localhost/*")
    assert application.redirect_uri_allowed("http://localhost/callback")

    application = ApplicationFactory(redirect_uris="*")
    assert application.redirect_uri_allowed("http://localhost/callback")

    # We do not handle wildcard in domains
    application = ApplicationFactory(redirect_uris="http://*.mydomain.com/callback")
    assert not application.redirect_uri_allowed("http://site1.mydomain.com/callback")


def test_logout(client):
    auth_url = reverse("oidc_overrides:authorize")
    user = UserFactory()
    client.force_login(user)
    response = client.get(auth_url)
    assert response.status_code == 400  # auth_url is missing all the arguments
    # TODO: Add a method to quickly to the oidc dance.

    assert get_user(client).is_authenticated
    logout_params = {"id_token_hint": 111}  # bad token
    # TODO: also try with existing token but expired
    response = client.get(add_url_params(reverse("oidc_overrides:logout"), logout_params))
    assertRedirects(response, "http://testserver/", fetch_redirect_response=False)
    assert not get_user(client).is_authenticated


def test_authorize_bad_oidc_params(client):
    # Application does not exist
    auth_url = reverse("oidc_overrides:authorize")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400


def test_authorize_not_authenticated(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    auth_url = reverse("oidc_overrides:authorize")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))
    assert client.session["next_url"] == auth_complete_url
    assert client.session[OIDCSessionMixin.OIDC_SESSION_KEY] == OIDC_PARAMS


def test_registrations_bad_oidc_params(client):
    # Application does not exist
    auth_url = reverse("oidc_overrides:registrations")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400


def test_registrations_not_authenticated(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    auth_url = reverse("oidc_overrides:registrations")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:registration"))
    assert client.session["next_url"] == auth_complete_url
    assert client.session[OIDCSessionMixin.OIDC_SESSION_KEY] == OIDC_PARAMS


def test_activation_bad_oidc_params(client):
    auth_url = reverse("oidc_overrides:activation")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400


def test_activation_missing_user_info(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    auth_url = reverse("oidc_overrides:activation")
    # Missing: email, firstname and lastname.
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    # The user is redirected to the activation view as the oidc parameters are valid
    assertRedirects(response, reverse("accounts:activation"), fetch_redirect_response=False)
    assert client.session["next_url"] == auth_complete_url
    assert client.session[OIDCSessionMixin.OIDC_SESSION_KEY] == OIDC_PARAMS

    response = client.get(response.url)
    # FIXME update the template
    assert response.status_code == 400


def test_activation_not_authenticated(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    auth_params = OIDC_PARAMS | {"email": "email", "firstname": "firstname", "lastname": "lastname"}
    auth_url = reverse("oidc_overrides:activation")
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:activation"))
    assert client.session["next_url"] == auth_complete_url
    assert client.session[OIDCSessionMixin.OIDC_SESSION_KEY] == auth_params
