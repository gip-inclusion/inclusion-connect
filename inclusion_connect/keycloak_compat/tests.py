import uuid

import jwt
import pytest
from django.contrib.auth import get_user
from django.contrib.auth.hashers import PBKDF2PasswordHasher
from django.urls import reverse
from oauth2_provider.models import get_access_token_model, get_id_token_model, get_refresh_token_model
from pytest_django.asserts import assertRedirects

from inclusion_connect.keycloak_compat.hashers import KeycloakPasswordHasher
from inclusion_connect.oidc_overrides.factories import DEFAULT_CLIENT_SECRET, ApplicationFactory, default_client_secret
from inclusion_connect.users.factories import UserFactory
from inclusion_connect.utils.urls import add_url_params, get_url_params


@pytest.mark.parametrize("realm", ["local", "Review_apps", "Demo", "inclusion-connect"])
def test_login(client, realm):
    application = ApplicationFactory()
    user = UserFactory()

    # Test AUTH endpoint when not authenticated
    auth_url = reverse(f"keycloak_compat_{realm}:authorize")
    auth_params = {
        "response_type": "code",
        "client_id": application.client_id,
        "redirect_uri": "http://localhost/callback",
        "scope": "openid profile email",
        "state": "state",
        "nonce": "nonce",
    }
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))
    assert client.session["next_url"] == auth_complete_url

    # Test AUTH endpoint when not authenticated and with bad params
    bad_auth_params = auth_params.copy()
    bad_auth_params["client_id"] = "toto"
    bad_auth_complete_url = add_url_params(auth_url, bad_auth_params)
    response = client.get(bad_auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400

    # Test AUTH endpoint when authenticated
    client.force_login(user)
    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(auth_params["redirect_uri"])
    auth_response_params = get_url_params(response.url)

    # Test TOKEN endpoint
    # FIXME it's recommanded to use basic auth here, maybe update our documentation ?
    token_data = {
        "client_id": application.client_id,
        "client_secret": DEFAULT_CLIENT_SECRET,
        "code": auth_response_params["code"],
        "grant_type": "authorization_code",
        "redirect_uri": "http://localhost/callback",
    }
    response = client.post(
        reverse(f"keycloak_compat_{realm}:token"),
        data=token_data,
    )
    token_json = response.json()
    id_token = token_json["id_token"]
    decoded_id_token = jwt.decode(
        id_token,
        key=default_client_secret(),
        algorithms=["HS256"],
        audience=application.client_id,
    )
    assert decoded_id_token["nonce"] == auth_params["nonce"]
    assert decoded_id_token["sub"] == str(user.pk)
    assert uuid.UUID(decoded_id_token["sub"]), "Sub should be an uuid"
    assert decoded_id_token["given_name"] == user.first_name
    assert decoded_id_token["family_name"] == user.last_name
    assert decoded_id_token["email"] == user.email

    # Test USER INFO endpoint
    response = client.get(
        reverse(f"keycloak_compat_{realm}:user-info"),
        HTTP_AUTHORIZATION=f"Bearer {token_json['access_token']}",
    )
    assert response.json() == {
        "sub": str(user.pk),
        "given_name": user.first_name,
        "family_name": user.last_name,
        "email": user.email,
    }

    # Test LOGOUT endpoint
    assert get_user(client).is_authenticated
    logout_params = {"id_token_hint": id_token}
    response = client.get(add_url_params(reverse(f"keycloak_compat_{realm}:logout"), logout_params))
    assert not get_user(client).is_authenticated
    assert get_id_token_model().objects.count() == 0
    assert get_access_token_model().objects.count() == 0
    assert get_refresh_token_model().objects.get().revoked is not None


@pytest.mark.parametrize("realm", ["local", "Review_apps", "Demo", "inclusion-connect"])
def test_registration(client, realm):
    application = ApplicationFactory()
    user = UserFactory()

    # Test REGISTRATIONS endpoint when not authenticated
    auth_url = reverse(f"keycloak_compat_{realm}:registrations")
    auth_params = {
        "response_type": "code",
        "client_id": application.client_id,
        "redirect_uri": "http://localhost/callback",
        "scope": "openid profile email",
        "state": "state",
        "nonce": "nonce",
    }
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:registration"))
    assert client.session["next_url"] == auth_complete_url

    # Test REGISTRATIONS endpoint when not authenticated and with bad oidc params
    bad_auth_params = auth_params.copy()
    bad_auth_params["client_id"] = "toto"
    bad_auth_complete_url = add_url_params(auth_url, bad_auth_params)
    response = client.get(bad_auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400

    # Test REGISTRATIONS endpoint when not authenticated with activation params
    activation_params = auth_params | {
        "login_hint": "email",
        "firstname": "John",
        "lastname": "Doe",
    }
    auth_complete_url = add_url_params(auth_url, activation_params)
    response = client.get(auth_complete_url, follow=True)
    assertRedirects(response, reverse("accounts:activation"))
    assert client.session["next_url"] == auth_complete_url

    # Test REGISTRATIONS endpoint when authenticated
    client.force_login(user)
    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(auth_params["redirect_uri"])
    auth_response_params = get_url_params(response.url)

    # Test TOKEN endpoint
    # FIXME it's recommanded to use basic auth here, maybe update our documentation ?
    token_data = {
        "client_id": application.client_id,
        "client_secret": DEFAULT_CLIENT_SECRET,
        "code": auth_response_params["code"],
        "grant_type": "authorization_code",
        "redirect_uri": "http://localhost/callback",
    }
    response = client.post(
        reverse(f"keycloak_compat_{realm}:token"),
        data=token_data,
    )
    token_json = response.json()
    id_token = token_json["id_token"]
    decoded_id_token = jwt.decode(
        id_token,
        key=default_client_secret(),
        algorithms=["HS256"],
        audience=application.client_id,
    )
    assert decoded_id_token["nonce"] == auth_params["nonce"]
    assert decoded_id_token["sub"] == str(user.pk)
    assert uuid.UUID(decoded_id_token["sub"]), "Sub should be an uuid"
    assert decoded_id_token["given_name"] == user.first_name
    assert decoded_id_token["family_name"] == user.last_name
    assert decoded_id_token["email"] == user.email

    # Test USER INFO endpoint
    response = client.get(
        reverse(f"keycloak_compat_{realm}:user-info"),
        HTTP_AUTHORIZATION=f"Bearer {token_json['access_token']}",
    )
    assert response.json() == {
        "sub": str(user.pk),
        "given_name": user.first_name,
        "family_name": user.last_name,
        "email": user.email,
    }

    # Test LOGOUT endpoint
    assert get_user(client).is_authenticated
    logout_params = {"id_token_hint": id_token}
    response = client.get(add_url_params(reverse(f"keycloak_compat_{realm}:logout"), logout_params))
    assert not get_user(client).is_authenticated
    assert get_id_token_model().objects.count() == 0
    assert get_access_token_model().objects.count() == 0
    assert get_refresh_token_model().objects.get().revoked is not None


def test_password_hasher(client):
    password = "RdaRfqP7Y89vy2"
    hashed_password = "$".join(
        [
            "keycloak-pbkdf2-sha256",
            "27500",
            "Td6XuopYK6JNfUnIlqYMOQ==",
            "ZXVC08Hf4jBOoYzVoNWYjQijsMC2oc/OUa9LciiIJ/1XHPF/qPiY1DqwLLDN2hYFmf/1kApkveD8/Pr7GVqjgw==",
        ]
    )
    user = UserFactory(password=hashed_password)
    assert user.password == hashed_password

    assert KeycloakPasswordHasher().verify(password=password, encoded=hashed_password)

    client.login(email=user.email, password=password)

    user.refresh_from_db()
    assert user.password != hashed_password
    assert PBKDF2PasswordHasher().verify(password, encoded=user.password)
