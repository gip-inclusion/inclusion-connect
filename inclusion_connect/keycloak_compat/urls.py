from django.urls import re_path
from oauth2_provider import views as oauth2_views

from ..accounts.views import EditUserInfoView
from ..oidc_overrides.views import AuthorizationView, LogoutView, RegistrationView
from . import views


app_name = "keycloak_compat"


urlpatterns = [
    re_path(
        r"^\.well-known/openid-configuration/$",
        oauth2_views.ConnectDiscoveryInfoView.as_view(),
        name="oidc-connect-discovery-info",
    ),
    re_path(r"^protocol/openid-connect/userinfo$", oauth2_views.UserInfoView.as_view(), name="user-info"),
    re_path(r"^protocol/openid-connect/auth$", AuthorizationView.as_view(), name="authorize"),
    re_path(r"^protocol/openid-connect/registrations$", RegistrationView.as_view(), name="registrations"),
    re_path(r"^protocol/openid-connect/token$", oauth2_views.TokenView.as_view(), name="token"),
    re_path(r"^protocol/openid-connect/logout$", LogoutView.as_view(), name="logout"),
    re_path(r"^account$", EditUserInfoView.as_view(), name="edit_user_info"),
    re_path(r"^login-actions/action-token$", views.ActionToken.as_view(), name="action-token"),
]
