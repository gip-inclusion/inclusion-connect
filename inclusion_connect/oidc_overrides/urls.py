from django.urls import re_path
from oauth2_provider import views as oauth2_views

from inclusion_connect.oidc_overrides import views


app_name = "oauth2_provider"

urlpatterns = [
    re_path(r"^authorize/", views.AuthorizationView.as_view(), name="authorize"),
    re_path(r"^register/", views.RegistrationView.as_view(), name="register"),
    re_path(r"^activate/", views.ActivationView.as_view(), name="activate"),
    re_path(r"^token/$", oauth2_views.TokenView.as_view(), name="token"),
    re_path(r"^revoke_token/$", oauth2_views.RevokeTokenView.as_view(), name="revoke-token"),
    re_path(r"^introspect/$", oauth2_views.IntrospectTokenView.as_view(), name="introspect"),
    # OIDC urls
    re_path(
        r"^\.well-known/openid-configuration/$",
        oauth2_views.ConnectDiscoveryInfoView.as_view(),
        name="oidc-connect-discovery-info",
    ),
    re_path(r"^\.well-known/jwks.json$", oauth2_views.JwksInfoView.as_view(), name="jwks-info"),
    re_path(r"^userinfo/$", oauth2_views.UserInfoView.as_view(), name="user-info"),
    re_path(r"^logout/", views.LogoutView.as_view(), name="logout"),
]
