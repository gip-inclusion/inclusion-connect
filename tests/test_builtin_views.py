from django.test import override_settings
from django.urls import reverse

from inclusion_connect.utils.urls import add_url_params
from tests.conftest import Client, parse_response_to_soup


def test_csrf_view(snapshot):
    with override_settings(MATOMO_BASE_URL="https://matomo.example.com", MATOMO_SITE_ID=1):
        client = Client(enforce_csrf_checks=True)
        response = client.post(
            add_url_params(
                reverse("accounts:login"),
                {
                    "referrer_uri": "http://go.back/there",
                    "other_parameter": "is_ignored",
                },
            ),
            {"username": "doesnot", "password": "matter"},
        )
        script_content = parse_response_to_soup(response)
        assert str(script_content) == snapshot
