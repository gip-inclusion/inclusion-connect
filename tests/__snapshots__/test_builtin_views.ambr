# serializer version: 1
# name: test_csrf_view
  '''
  
  
  
  <!DOCTYPE HTML>
  <html lang="fr">
      <head>
          <meta charset="utf-8">
          <title>Inclusion Connect</title>
          <meta name="robots" content="noindex, nofollow">
          <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  
          <link rel="icon" href="/static/img/illustration-bg-ic.svg" type="image/svg+xml">
          <link rel="icon" href="/static/img/favicon.ico" type="image/ico">
          <link rel="stylesheet" href="/static/vendor/theme-inclusion/stylesheets/app.css" type="text/css">
          <link rel="stylesheet" href="/static/css/inclusion_connect.css" type="text/css">
  
          
      </head>
      <body>
          
  
  
  <header class="s-header s-header--emploi sticky-top" role="banner" id="header">
      <section class="s-header__container container">
          <div class="s-header__row row">
              <div class="s-header__col s-header__col--logo col-auto d-flex align-items-center pr-0 pr-md-3">
                  <img src="/static/vendor/theme-inclusion/images/logo-ministere-emploi.svg" alt="Ministère du travail, du plein emploi et de l'insertion" />
              </div>
              <div class="s-header__col s-header__col--logo col-auto d-flex align-items-center px-0 px-md-3">
                  <img src="/static/vendor/theme-inclusion/images/logo-inclusion-connect.svg" alt="Inclusion Connect" />
              </div>
          </div>
      </section>
  </header>
  
  
          <main id="main" role="main" class="s-main">
  
              <section class="s-section pt-lg-3">
                  <div class="container">
                      <div class="row no-gutters bg-white">
                          
      <div class="col-12 col-lg-6 bg-lg-white">
          <div class="bg-white p-3 py-lg-7 px-lg-8">
              <div class="container container-messages">
                  
                      
  
  
                  
              </div>
  
              
      <div class="alert alert-danger">
          <p>
              La requête a été interrompue pour votre sécurité (mécanisme <abbr>CSRF</abbr>).
              Peut-être vous êtes-vous connecté depuis un autre onglet ?
          </p>
          <p class="mb-0">
              Pour continuer à utiliser Inclusion Connect,
              <a href="/accounts/my-account/?redirect_uri=http%3A%2F%2Fgo.back%2Fthere">retournez sur la page de gestion de votre compte</a>.
          </p>
      </div>
  
          </div>
      </div>
  
                          
      <div class="d-none d-lg-flex col-lg-6 p-lg-6 justify-content-center align-items-center bg-light">
          <img class="img-fluid" src="/static/img/illustration-ic.svg" alt="" loading="lazy">
      </div>
  
                      </div>
                  </div>
              </section>
  
  
          </main>
  
          
  
  
  <footer class="s-footer" role="contentinfo" id="footer">
      <section class="s-footer-nav">
          <div class="s-footer-nav__container container">
              <div class="s-footer-nav__row row">
                  <div class="s-footer-nav__col col-6 d-flex align-items-center">
                      <div class="s-footer-nav__logo">
                          <img src="/static/vendor/theme-inclusion/images/logo-republique-francaise.svg" alt="Rébublique Française" class="img-fluid">
                      </div>
                  </div>
                  <div class="s-footer-nav__col col-12 col-lg-6 d-flex align-items-center small">
                      <div>
                          <p class="mb-3">
                              Inclusion Connect est un Single Sign-On (SSO). Autrement dit c'est un service d'authentification qui permet à un utilisateur d'utiliser les mêmes informations d'identification (email et mot de passe) pour accéder à plusieurs applications.
                          </p>
                          <ul>
                              <li>
                                  <a target="_blank" href="https://legifrance.gouv.fr">
                                      legifrance.gouv.fr<i class="ri-external-link-line ml-1 ri-lg" aria-hidden="true"></i>
                                  </a>
                              </li>
                              <li>
                                  <a target="_blank" href="https://gouvernement.fr">
                                      gouvernement.fr<i class="ri-external-link-line ml-1 ri-lg" aria-hidden="true"></i>
                                  </a>
                              </li>
                              <li>
                                  <a target="_blank" href="https://service-public.fr">
                                      service-public.fr<i class="ri-external-link-line ml-1 ri-lg" aria-hidden="true"></i>
                                  </a>
                              </li>
                              <li>
                                  <a target="_blank" href="https://data.gouv.fr">
                                      data.gouv.fr<i class="ri-external-link-line ml-1 ri-lg" aria-hidden="true"></i>
                                  </a>
                              </li>
                          </ul>
                      </div>
                  </div>
                  <div class="fr-footer__content col-12 col-lg-6 d-flex align-items-center">
                      <p class="fr-footer__content-desc"></p>
                      <ul class="fr-footer__content-list">
                      </ul>
                  </div>
              </div>
          </div>
      </section>
  
      <section class="s-footer-legal">
          <div class="s-footer-legal__container container">
              <div class="s-footer-legal__row row">
                  <div class="s-footer-legal__col col-xs-12 col-lg-8">
                      <ul>
                          <li>
                              <a href="/static/terms/Politique%20de%20confidentialite%CC%81_20230512.pdf" target="_blank">
                                  Politique de confidentialité<i class="ri-external-link-line ml-1 ri-lg" aria-hidden="true"></i>
                              </a>
                          </li>
                          <li>
                              <a href="/static/terms/CGU_20230302.pdf" target="_blank">
                                  CGU<i class="ri-external-link-line ml-1 ri-lg" aria-hidden="true"></i>
                              </a>
                          </li>
                          <li>
                              <a href="/static/terms/Mentions%20l%C3%A9gales_20230302.pdf" target="_blank">
                                  Mentions légales<i class="ri-external-link-line ml-1 ri-lg" aria-hidden="true"></i>
                              </a>
                          </li>
                      </ul>
                  </div>
              </div>
          </div>
      </section>
  </footer>
  
  
          
          
          
  
          <script src="/static/vendor/jquery-3.6.4/jquery-3.6.4.min.js"></script>
          <script src="/static/vendor/bootstrap-4.3.1/popper.min.js"></script>
          <script src="/static/vendor/bootstrap-4.3.1/bootstrap.min.js"></script>
          <script src="/static/vendor/theme-inclusion/javascripts/app.js"></script>
  
          
              <script src="/static/js/index.js"></script>
          
      </body>
  </html>
  
  '''
# ---
